### PR TITLE
Add U and Phase gate

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Adds `U/CU` and `Phase/CPhase` gates. 

### Details and comments

The phase gate is equivalent to U1, the U gate equivalent to U3. Unlike CU3 which contains 3 parameters, CU has a total of 4 parameters, where the new one is used to account for a possible phase of the controlled-U gate.

The U gate is called `'u'`, the phase gate called `'P'` (which is also how they are displayed in plots).

U1/U2/U3 are not yet deprecated. Since this will probably touch a lot of code where references to `'u1'` etc. are hardcoded, this will be addressed in a separate issue.

Partially addresses #4106, but doesn't close it since U1/U2/U3 are not removed.

### CODE REVIEW

File: qiskit/circuit/library/standard_gates/equivalence_library.py [hunk 1]
```diff
@@ -355,14 +385,62 @@
...
+theta = Parameter('theta')
+phi = Parameter('phi')
+lam = Parameter('lam')
+gamma = Parameter('gamma')
+def_cu = QuantumCircuit(q)
+def_cu.p(gamma, 0)
+def_cu.p((lam + phi) / 2, 0)
+def_cu.p((lam - phi) / 2, 1)
+def_cu.cx(0, 1)
+def_cu.u3(-theta / 2, 0, -(phi + lam) / 2, 1)  # TODO use QuantumCircuit.u but fails a test
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/equivalence_library.py:422]: @kdk I don't know what's going on there, it seems like a parameter binding issue like the one we discussed before where we had a CPhase whose parameter wasn't properly bound.

Comment (kdk) [reply qiskit/circuit/library/standard_gates/equivalence_library.py:422]: I think what's missing is a cast of params to `float` in `CUGate.to_matrix` (similar to https://github.com/Qiskit/qiskit-terra/blob/8f717d9/qiskit/circuit/library/standard_gates/u3.py#L95 ).

File: qiskit/circuit/library/standard_gates/equivalence_library.py [hunk 2]
```diff
@@ -100,6 +104,32 @@
             def_ms.append(RXXGate(theta), [q[i], q[j]])
     _sel.add_equivalence(MSGate(num_qubits, theta), def_ms)
 
+# PhaseGate
+
+q = QuantumRegister(1, 'q')
+theta = Parameter('theta')
+phase_to_u1 = QuantumCircuit(q)
+phase_to_u1.u1(theta, 0)
+_sel.add_equivalence(PhaseGate(theta), phase_to_u1)
```
Comment (kdk) [inline qiskit/circuit/library/standard_gates/equivalence_library.py:113]: Can you also add the reverse mapping `u1_to_phase`? (Same for `u3_to_u`)

File: qiskit/circuit/library/standard_gates/equivalence_library.py (hunk 1 above)
Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/equivalence_library.py]: that fixed it! done in dfb2af9

File: qiskit/circuit/library/standard_gates/equivalence_library.py (hunk 2 above)
Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/equivalence_library.py:113]: done in 3140040

File: qiskit/circuit/library/standard_gates/u.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+        .. math::
+
+            CU(\theta, \phi, \lambda)\ q_1, q_0 =
+                |0\rangle\langle 0| \otimes I +
+                e^{i\gamma}|1\rangle\langle 1| \otimes U3(\theta,\phi,\lambda) =
+                \begin{pmatrix}
+                    1 & 0 & 0                             & 0 \\
+                    0 & 1 & 0                             & 0 \\
+                    0 & 0 & e^{i\gamma} \cos(\th)         & -e^{i(\gamma + \lambda)}\sin(\th) \\
```
Comment (kdk) [inline qiskit/circuit/library/standard_gates/u.py:169]:
```suggestion
                    0 & 0 & e^{i\gamma} \cos(\th/2)         & -e^{i(\gamma + \lambda)}\sin(\th/2) \\
```

File: qiskit/circuit/library/standard_gates/u.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+                    0 & 0 & e^{i(\gamma + \phi)}\sin(\th) & e^{i(\gamma + \phi+\lambda)\cos(\th)}
```
Comment (kdk) [inline qiskit/circuit/library/standard_gates/u.py:170]:
```suggestion
                    0 & 0 & e^{i(\gamma + \phi)}\sin(\th/2) & e^{i(\gamma + \phi+\lambda)\cos(\th/2)}
```

File: qiskit/circuit/library/standard_gates/u.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    def __init__(self, theta, phi, lam, gamma, label=None, ctrl_state=None):
+        """Create new CU gate."""
+        super().__init__('cu', 2, [theta, phi, lam, gamma], num_ctrl_qubits=1,
+                         label=label, ctrl_state=ctrl_state)
+        self.base_gate = UGate(theta, phi, lam)
+
+    def _define(self):
+        """
+        gate cu(theta,phi,lambda,gamma) c, t
+        { cphase(gamma) c,t;
```
Comment (kdk) [inline qiskit/circuit/library/standard_gates/u.py:183]:
```suggestion
        { phase(gamma) c;
```

File: qiskit/qasm/libs/qelib1.inc [hunk 6]
```diff
@@ -122,6 +134,15 @@ gate cu3(theta,phi,lambda) c, t
   cx c,t;
   u3(theta/2,phi,0) t;
 }
+gate cu(theta,phi,lambda,gamma) c, t
+{ cp(gamma) c,t;
```
Comment (kdk) [inline qiskit/qasm/libs/qelib1.inc:138]:
```suggestion
{ p(gamma) c;
```

File: qiskit/circuit/library/standard_gates/p.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+            return gate
+        return super().control(num_ctrl_qubits=num_ctrl_qubits, label=label, ctrl_state=ctrl_state)
+
+    def inverse(self):
+        r"""Return inverted CPhase gate (:math:`CPhase(\lambda){\dagger} = CPhase(-\lambda)`)"""
+        return CPhaseGate(-self.params[0])
+
+    def to_matrix(self):
+        """Return a numpy.array for the CPhase gate."""
+        eith = numpy.exp(1j * self.params[0])
```
[the remark this answers is not in the document]
Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/p.py:205]: Good question, I don't know. Maybe that was only prone to failure if the gates were used in a nested order or so...

File: qiskit/circuit/library/standard_gates/u.py (hunk 3 above)
Comment (Cryoris) [reply qiskit/circuit/library/standard_gates/u.py:169]: \th is already \theta/2, see L135

File: qiskit/circuit/library/standard_gates/u.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        .. math::
+
+            CU(\theta, \phi, \lambda)\ q_1, q_0 =
+                |0\rangle\langle 0| \otimes I +
+                e^{i\gamma}|1\rangle\langle 1| \otimes U3(\theta,\phi,\lambda) =
+                \begin{pmatrix}
+                    1 & 0 & 0                             & 0 \\
+                    0 & 1 & 0                             & 0 \\
+                    0 & 0 & e^{i\gamma} \cos(\th)         & -e^{i(\gamma + \lambda)}\sin(\th) \\
+                    0 & 0 & e^{i(\gamma + \phi)}\sin(\th) & e^{i(\gamma + \phi+\lambda)\cos(\th)}
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/u.py:170]:
```suggestion
                    0 & 0 & e^{i(\gamma + \phi)}\sin(\th) & e^{i(\gamma + \phi+\lambda)}\cos(\th)
```

File: qiskit/circuit/library/standard_gates/u.py [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+        \newcommand{\th}{\frac{\theta}{2}}
+
+        CU(\theta, \phi, \lambda)\ q_0, q_1 =
+            I \otimes |0\rangle\langle 0| +
+            e^{i\gamma} U(\theta,\phi,\lambda) \otimes |1\rangle\langle 1| =
+            \begin{pmatrix}
+                1 & 0                           & 0 & 0 \\
+                0 & e^{i\gamma}\cos(\th)        & 0 & -e^{i(\gamma + \lambda)}\sin(\th) \\
+                0 & 0                           & 1 & 0 \\
+                0 & e^{i(\gamma+\phi)}\sin(\th) & 0 & e^{i(\gamma+\phi+\lambda)\cos(\th)}
```
Comment (Cryoris) [inline qiskit/circuit/library/standard_gates/u.py:144]:
```suggestion
                0 & e^{i(\gamma+\phi)}\sin(\th) & 0 & e^{i(\gamma+\phi+\lambda)}\cos(\th)
```

**Pull request merged**